### DISCUSSION TITLE
After adding post-processing to the model output, repackage and export the new model

### DISCUSSION THREAD

**Discussion by canhe173** [ultralytics/ultralytics#16182] [Q&A]
Generally, for segmentation models in yolov8, it needs to post-process the model output before we can get information like boxes, classes, scores. Therefore, I want to repackage this post-processing part behind the original output of the model, and finally export a new model containing 4 outputs, which includes boxes, classes, scores, and masks.

Below is my code, it cannot export model with new outputs and it still export the original model. Maybe I need some help.

```python
import torch
import torchvision
from ultralytics import YOLO

class YoloModelWithPostProcess(YOLO):
    def __init__(self, model_path):
        super(YoloModelWithPostProcess, self).__init__(model_path)
        self.YOLO = YOLO(model_path)

    def forward(self, x):
        yolo_output = self.YOLO(x)
        boxes, confidences, class_probs, masks = self.postprocess(yolo_output)
        return boxes, confidences, class_probs, masks
    
    def postprocess(self, preds, conf_threshold=0.025, iou_threshold = 0.045, nm=32):

        x, protos = preds[0], preds[1]
        x = x.permute(0, 2, 1)
        # Predictions filtering by conf-threshold
        x = x[torch.amax(x[..., 4:-nm], axis=-1) > conf_threshold]
        # Create a new matrix which merge these(box, score, cls, nm) into one
        scores = torch.amax(x[..., 4:-nm], dim=-1)
        class_indices = torch.argmax(x[..., 4:-nm], dim=-1)
        x = torch.cat([x[..., :4], scores.unsqueeze(-1), class_indices.unsqueeze(-1), x[..., -nm:]], dim=-1)
        # NMS filtering
        x = x[torchvision.ops.nms(x[:, :4], x[:, 4], iou_threshold).numpy()]
        # Decode and return
        if len(x) > 0:
            # Bounding boxes format change: cxcywh -> xyxy
            x[..., [0, 1]] -= x[..., [2, 3]] / 2
            x[..., [2, 3]] += x[..., [0, 1]]
   
            boxes_output = x[..., :4]
            confidences_output = x[..., 4:5]
            class_probs_output = x[..., 5:6]
            # Process masks
            c, mh, mw = protos[2][0].shape
            masks = torch.matmul(x[:, 6:], protos.view(c, -1)).view((-1, mh, mw)).permute(1, 2, 0)  # HWN
            masks_output = torch.einsum("HWN -> NHW", masks)  # HWN -> NHW
        
            return boxes_output, confidences_output, class_probs_output, masks_output
        

model_with_postprocess = YoloModelWithPostProcess('yolov8n-seg.pt')
model_with_postprocess.export(format="onnx")
```

**Comment by glenn-jocher** [comment]
To export a model with custom outputs, you'll need to modify the model's forward method and ensure the export process captures these changes. Consider using a custom wrapper or script to handle the post-processing and export the model with the new outputs. If issues persist, reviewing the ONNX export documentation for custom operations might help.

**Reply by canhe173** [reply]
Thanks for your reply. What do you mean by using a custom wrapper? Is it similar to the code I wrote above? Moreover, you recommend me to modify the model's forward, but to be honest, I haven't found which file the model forward is in. Could you give me some more specific instructions?

**Reply by glenn-jocher** [reply]
@canhe173 a custom wrapper involves creating a new class or function to handle post-processing, similar to your approach. To modify the model's forward method, you can extend the YOLO class and override the forward method directly in your script. If you need more details, reviewing the YOLOv8 source code in the Ultralytics repository might help.
